src/utils.ts

```diff
@@ -25,68 +24,100 @@ import { PaymentRequest } from './model/PaymentRequest.js';
 				return curr + acc;
 			}, 0);
 	}
-	const sortedKeyAmounts: Array<number> = Object.keys(keyset)
-		.map((k) => parseInt(k))
-		.sort((a, b) => b - a);
-	sortedKeyAmounts.forEach((amt) => {
+	const sortedKeyAmounts = getKeysetAmounts(keyset);
+	sortedKeyAmounts.forEach((amt: number) => {
 		const q = Math.floor(value / amt);
 		for (let i = 0; i < q; ++i) chunks.push(amt);
 		value %= amt;
 	});
-	return chunks.sort((a, b) => (isDesc ? b - a : a - b));
+	return chunks.sort((a, b) => (order === 'desc' ? b - a : a - b));
+}
+
+function getKeepAmounts(
+	proofsWeHave: Array<Proof>,
+	amountToKeep: number,
+	keys: Keys,
+	targetCount: number
+): Array<number> {
+	// determines amounts we need to reach the targetCount for each amount based on the amounts of the proofs we have
+	// it tries to select amounts so that the proofs we have and the proofs we want reach the targetCount
+	const amountsWeWant: Array<number> = [];
+	const amountsWeHave = proofsWeHave.map((p: Proof) => p.amount);
+	const sortedKeyAmounts = getKeysetAmounts(keys, 'asc');
+	sortedKeyAmounts.forEach((amt) => {
+		const countWeHave = amountsWeHave.filter((a) => a === amt).length;
+		const countWeWant = Math.floor(targetCount - countWeHave);
+		for (let i = 0; i < countWeWant; ++i) {
+			if (amountsWeWant.reduce((a, b) => a + b, 0) + amt > amountToKeep) {
+				break;
+			}
+			amountsWeWant.push(amt);
+		}
+	});
+	// use splitAmount to fill the rest between the sum of amountsWeHave and amountToKeep
+	const amountDiff = amountToKeep - amountsWeWant.reduce((a, b) => a + b, 0);
+	if (amountDiff) {
+		const remainingAmounts = splitAmount(amountDiff, keys);
+		remainingAmounts.forEach((amt: number) => {
+			amountsWeWant.push(amt);
+		});
+	}
+	const sortedAmountsWeWant = amountsWeWant.sort((a, b) => a - b);
+	// console.log(`# getKeepAmounts: amountToKeep: ${amountToKeep}`);
+	// console.log(`# getKeepAmounts: amountsWeHave: ${amountsWeHave}`);
+	// console.log(`# getKeepAmounts: amountsWeWant: ${sortedAmountsWeWant}`);
+	return sortedAmountsWeWant;
 }
 
-/*
-function isPowerOfTwo(number: number) {
-	return number && !(number & (number - 1));
+// function isPowerOfTwo(number: number) {
+// 	return number && !(number & (number - 1));
+// }
+function getKeysetAmounts(keyset: Keys, order = 'desc'): Array<number> {
+	if (order == 'desc') {
+		return Object.keys(keyset)
+			.map((k: string) => parseInt(k))
+			.sort((a: number, b: number) => b - a);
+	}
+	return Object.keys(keyset)
+		.map((k: string) => parseInt(k))
+		.sort((a: number, b: number) => a - b);
 }
-*/
 
-function hasCorrespondingKey(amount: number, keyset: Keys): boolean {
+function hasCorrespondingKey(amount: number, keyset: Keys) {
 	return amount in keyset;
 }
 
-function getPreference(
-	amount: number,
-	keyset: Keys,
-	preferredAmounts: Array<AmountPreference>
-): Array<number> {
+function getPreference(amount: number, keyset: Keys, split: Array<number>): Array<number> {
 	const chunks: Array<number> = [];
```

Comment:
is this function still needed? it's only used in `splitAmount` and seems all that it does is make sure all the split amounts have a corresponding key.
Also, the `amount` parameter is not used any more in this function